File: languages/javascript/check.js
#! /usr/bin/env node

/**
   check.js
   ====================================================
   CREATED: 2018-05-14
   UPDATED: 2018-06-06
   VERSION: 1.0.0
   AUTHOR:  wlharvey4
   ABOUT:   Test script for JavaScript code challenges.
   USAGE:   ./check <code-challenge>
   NOTES:
   ----------------------------------------------------
 */

// utility libraries
const assert = require('assert').strict;
const fs     = require('fs');
const path   = require('path');

// constants
const OK = 'ok';
const FAILED = 'failed';
const USAGE  = 'USAGE: ./check <code-challenge>'; 
const CHALLENGES = '/challenges';

let cc;  // the name of the code challenge, from the CL
try {
  cc = process.argv[2];
  if (!cc) throw new Error("ERROR: Need a code challenge name");
} catch(e) {
  console.error(e.message);
  console.log(USAGE);
  process.exit(-1);
}

// construct paths to resources
const challengesDir = path.join (
  /* absolute path to `challenges' directory
     from languages/<language> directory */
  path.format({root: path.dirname(path.dirname(path.dirname(module.filename)))}),
  CHALLENGES
);
// challenges/<code-challenge>/
const ccDir = path.format({
  dir: challengesDir,
  name: cc
});
// JSON file holding test data for code challenge
const ccJSON = path.format({ // path to JSON data
  dir: ccDir,
  name: cc,
  ext: '.json'
});

// load and parse the test data into a JS object
const jsonData = JSON.parse(fs.readFileSync(ccJSON, {encoding: 'utf-8'}));

const ccLangDir = path.format({
  dir: ccDir,
  name: 'javascript'
});
const ccJs = path.format({ // path to code challenge
  dir: ccLangDir,
  name: cc,
  ext: '.js'
});

// the code challenge
const fn = require(ccJs);

const assertError = (params, result, expected) => {
  console.error('----------------------------------------------------');
  console.error('Params:  ', params);
  console.error('Expected:', expected);
  console.error('Received:', result);
  console.error('----------------------------------------------------\n');
}

// the test runner; uses assert.deepStrictEqual
results = jsonData.reduce((results, datum) => {
  const result = fn(datum.params);
  try {
    assert.deepStrictEqual(result, datum.expected);
    results[OK]++;
  } catch(e) {
    assertError(datum.params, result, datum.expected);
    results[FAILED]++;
  }
  return results;
}, {ok: 0, failed: 0})

console.error('====================================================');
console.log(`Attempted:\t${results.ok + results.failed}`);
console.log(`Ok:\t\t${results.ok}`);
console.log(`Failed:\t\t${results.failed}`);
console.log('====================================================');

File: languages/perl/check.pl
#! /usr/bin/env perl5

# languages/perl/check.pl
# =============================================================================
# CREATED: 2018-05-19
# UPDATED: 2018-10-24
# VERSION: 3.1.1
# AUTHOR : wlharvey4
# ABOUT  : Test script for perl Perl code challenges
# USAGE  : ./check <code-challenge>
# NOTES: : While coding the challenges in OO paradigm might be cumbersome in
#	   Perl, yet there are real benefits to doing9 so, i.e., no aliasing of
#	   typeglobs is necessary, and therefore no need for 'no strict refs',
#	   nor for a global variable and checking for equality is easy and
#	   natural.
# CHANGE-LOG:
# .............................................................................
# v1.1.0 2018-07-04
# -- added `eval' (i.e., `try') around check for $e->isa Boolean because a null
#    $e throws an error.
# .............................................................................
# v1.1.1 2018-09-04
# -- changed the interpreter from perl to perl5
# .............................................................................
# v2.0.0 2018-09-07
# -- Added package names to each perl module; updated routine to alias the
#    package name along with the code challenge name
# .............................................................................
# v2.1.0 2018-09-08
# -- Added `bigint' pragma to `fibonacci' code challenge, which broke the
#    eq_deeply' check because the return value is a Math::BigInt object; so now
#    check to see if the return value is a Math::BigInt object, and if so, then
#    convert the `expected' value to a Math::BigInt object also
# .............................................................................
# v3.0.0 2018-10-24T0015
# -- Refactored to interface with OO code challenge; works with strCount()
# .............................................................................
# v3.1.0 2018-10-24T00:45
# -- Made further refactorings; gave variables better names; reduced comments;
# -- removed all global variables
# .............................................................................
# v3.1.1 2018-10-24T21:40
# -- refactored method output() to method cc(); nothing needed to be changed
#    to work with complete fizzbuzz refactoring into Perl OO style.
# -----------------------------------------------------------------------------

# pragmas
use strict;
use warnings;
use v5.16;
use boolean;	           # gives access to Perl boolean values

# Utility libraries
use Cwd 'abs_path';        # provides abs_path() function
use File::Spec::Functions; # provides &catfile() for portable path creation
use Data::Printer output => 'stdout'; # provides &p() (pretty-printing of data)
use JSON;                  # provides &decode_json()

# CONSTANTS
my $USAGE      = qq{USAGE: ./check <code-challenge>\n};
my $CHALLENGES = abs_path("../../challenges");
my $JSONEXT    = ".json";
my $PERL       = "perl";
my $PERLPM     = ".pm";

# Variables
my $cc;         # the name of the code challenge from the CL invocation
my $ccPM;       # absolute path to the code challenge module
my $ccJSON;     # absolute path to the code challenge JSON data file
my $jsonData;   # holds the JSON data in memory
my $packg;      # code challenge Package name
my ($i, $w);    # $i :=  test number; $w := number of tests found wrong

# get the Code Challenge name from the command-line argument and create the
# package name
$cc = $ARGV[0] || do {
    print $USAGE; print qq{ERROR: missing <code-challenge>\n}; exit 1;
};
$packg = ucfirst $cc;

# construct paths to resources
$ccJSON = catfile($CHALLENGES, $cc, $cc.$JSONEXT);
$ccPM   = catfile($CHALLENGES, $cc, $PERL, $cc.$PERLPM);
unless (-e $ccPM) {
    print $USAGE;
    print qq{ERROR: $ccPM does not exist\n};
    exit 1;
}

require $ccPM;     # load the code challenge module
$jsonData = eval { # load the JSON data into an Array ref
    local $/;
    open JSON, '<', $ccJSON;
    decode_json(<JSON>);
} || die $@;

($i, $w) = (0, 0);
for my $test (@$jsonData) { # iterate over the JSON tests
    $i++;
    my $params   = $test->{params};
    my $expected = $test->{expected};
    # need to convert JSON boolean to Perl boolean because
    # a null $e will throw an error
    if (eval { $expected->isa("JSON::PP::Boolean") }) {  
	$expected = boolean($expected)			  
    }

    my $result = $packg->$cc($params); # when using OO Perl, there is no need 
                                       # to alias type globs

    # if a code challenge uses the `bigint' pragma (e.g., `fibonacci'), need to
    # convert the `expected' value to a Math::BigInt object in order to
    # compare; don't want to simply load the `bigint' pragma because then all
    # math operations will be converted into Math::BigInt ones
    if ($result->cc()->isa("Math::BigInt")) {
	$expected = Math::BigInt->new($expected);
    }
    assertError($params, $result, $expected)  unless $result->eq($expected);
}

my $total = scalar @$jsonData;
printf(qq{====================================================
Attempted:\t%d
Ok:\t\t%d
Failed:\t\t%d
====================================================
}, $total, $total-$w, $w);

sub assertError {
    my ($p, $r, $e) = @_;
    say("----------------------------------------------------");
    say("Test [$i]");
    print("Params:   => ");
    p($p);
    print("Expected: => ");
    p($e);
    print("Received: => ");
    p($r->cc());
    print("Package   => ");
    p($r);
    print("----------------------------------------------------\n");
    $w++;
}
